File: app-launch-toolkit/scripts/check-site.mjs
import { readFile, readdir, stat } from 'node:fs/promises';
import { join, extname } from 'node:path';
import { tools } from '../src/data/tools.mjs';
import { guides } from '../src/data/guides.mjs';
import { templates } from '../src/data/templates.mjs';
import { pages } from '../src/data/pages.mjs';

const dist = 'dist';
const expectedRoutes = [
  '/',
  '/tools/',
  ...tools.map((tool) => `/tools/${tool.slug}/`),
  '/guides/',
  ...guides.map((guide) => `/guides/${guide.slug}/`),
  '/templates/',
  ...templates.map((template) => `/templates/${template.slug}/`),
  ...pages.map((page) => `/${page.slug}/`),
  '/404.html'
];

const errors = [];
const htmlFiles = [];

async function exists(path) {
  try {
    await stat(path);
    return true;
  } catch {
    return false;
  }
}

async function walk(dir) {
  for (const entry of await readdir(dir)) {
    const path = join(dir, entry);
    const info = await stat(path);
    if (info.isDirectory()) await walk(path);
    else if (extname(path) === '.html') htmlFiles.push(path);
  }
}

function routeFile(route) {
  if (route === '/') return join(dist, 'index.html');
  if (route.endsWith('.html')) return join(dist, route);
  return join(dist, route, 'index.html');
}

for (const route of expectedRoutes) {
  if (!(await exists(routeFile(route)))) errors.push(`Missing route: ${route}`);
}
if (!(await exists(join(dist, 'sitemap.xml')))) errors.push('Missing sitemap.xml');
if (!(await exists(join(dist, 'robots.txt')))) errors.push('Missing robots.txt');

await walk(dist);

const seenTitles = new Map();
for (const file of htmlFiles) {
  const html = await readFile(file, 'utf8');
  const route = file.replace(/^dist/, '').replace(/\/index\.html$/, '/').replace(/\/404\.html$/, '/404.html') || '/';
  const title = html.match(/<title>(.*?)<\/title>/)?.[1];
  const description = html.match(/<meta name="description" content="(.*?)">/)?.[1];
  if (!title) errors.push(`${file}: missing title`);
  if (!description || description.length < 40) errors.push(`${file}: missing/short meta description`);
  if (!html.includes('rel="canonical"')) errors.push(`${file}: missing canonical`);
  if (!html.includes('property="og:title"')) errors.push(`${file}: missing Open Graph tags`);
  if (!html.includes('name="twitter:card"')) errors.push(`${file}: missing Twitter card`);
  if (!html.includes('<h1')) errors.push(`${file}: missing h1`);
  if (title) {
    if (seenTitles.has(title)) errors.push(`${file}: duplicate title also in ${seenTitles.get(title)}`);
    seenTitles.set(title, file);
  }
  const links = [...html.matchAll(/href="(\/[^"#?]*)/g)].map((match) => match[1]);
  for (const link of links) {
    if (link.startsWith('/assets/')) continue;
    const normalized = link.endsWith('/') || link.endsWith('.html') ? link : `${link}/`;
    if (!expectedRoutes.includes(normalized) && !['/sitemap.xml', '/robots.txt'].includes(normalized)) {
      errors.push(`${file}: broken internal link ${link}`);
    }
  }
}

const sitemap = await readFile(join(dist, 'sitemap.xml'), 'utf8');
for (const route of expectedRoutes.filter((route) => route !== '/404.html')) {
  if (!sitemap.includes(route === '/' ? 'https://applaunchtoolkit.com/' : `https://applaunchtoolkit.com${route}`)) {
    errors.push(`sitemap missing ${route}`);
  }
}

if (errors.length) {
  console.error(errors.join('\n'));
  process.exit(1);
}
console.log(`Site check passed: ${expectedRoutes.length} expected routes, ${htmlFiles.length} HTML files.`);

File: app-launch-toolkit/scripts/check_site.py
from pathlib import Path
import re

expected = ['/']
expected += ['/tools/'] + [f'/tools/{slug}/' for slug in [
'app-store-screenshot-size-checker','app-store-screenshot-resizer','app-store-title-generator','app-store-subtitle-generator','app-store-keyword-counter','whats-new-generator','app-review-notes-generator','privacy-policy-generator','terms-of-use-generator','aso-keyword-generator','app-localization-checklist','subscription-price-calculator','app-store-listing-checker']]
expected += ['/guides/'] + [f'/guides/{slug}/' for slug in ['app-store-screenshot-sizes','how-to-write-app-store-subtitle','app-store-keyword-field-guide','app-privacy-policy-url-guide','app-store-review-notes-guide','app-store-localization-checklist']]
expected += ['/templates/'] + [f'/templates/{slug}/' for slug in ['app-review-notes-template','app-privacy-policy-template','app-store-submission-checklist','app-localization-checklist-template']]
expected += ['/about/','/contact/','/privacy-policy/','/terms/','/404.html']

dist = Path('dist')
errors = []

def file_for(route):
    if route == '/': return dist / 'index.html'
    if route.endswith('.html'): return dist / route.lstrip('/')
    return dist / route.lstrip('/') / 'index.html'

for route in expected:
    if not file_for(route).exists():
        errors.append(f'Missing route: {route}')

if not (dist / 'sitemap.xml').exists(): errors.append('Missing sitemap.xml')
if not (dist / 'robots.txt').exists(): errors.append('Missing robots.txt')

titles = {}
html_files = list(dist.rglob('*.html')) if dist.exists() else []
for file in html_files:
    html = file.read_text(encoding='utf-8')
    title = re.search(r'<title>(.*?)</title>', html)
    desc = re.search(r'<meta name="description" content="(.*?)">', html)
    if not title: errors.append(f'{file}: missing title')
    elif title.group(1) in titles: errors.append(f'{file}: duplicate title with {titles[title.group(1)]}')
    elif len(title.group(1)) < 10: errors.append(f'{file}: short title')
    else: titles[title.group(1)] = file
    if not desc or len(desc.group(1)) < 40: errors.append(f'{file}: missing/short description')
    for needle in ['rel="canonical"','property="og:title"','name="twitter:card"','<h1']:
        if needle not in html: errors.append(f'{file}: missing {needle}')
    for link in re.findall(r'href="(/[^"#?]*)', html):
        if link.startswith('/assets/') or link in ['/sitemap.xml','/robots.txt']:
            continue
        normalized = link if link.endswith('/') or link.endswith('.html') else link + '/'
        if normalized not in expected:
            errors.append(f'{file}: broken internal link {link}')

sitemap = (dist / 'sitemap.xml').read_text(encoding='utf-8') if (dist / 'sitemap.xml').exists() else ''
for route in expected:
    if route == '/404.html': continue
    url = 'https://applaunchtoolkit.com/' if route == '/' else 'https://applaunchtoolkit.com' + route
    if url not in sitemap:
        errors.append(f'sitemap missing {route}')

if errors:
    print('\n'.join(errors))
    raise SystemExit(1)
print(f'Site check passed: {len(expected)} expected routes, {len(html_files)} HTML files.')
